File: src/format/javadoc/nesting-stack.ts
// Port of google-java-format core/.../java/javadoc/NestingStack.java.
//
// A generic nesting stack (used by the lexer for HTML/code/table contexts) and
// an Int variant (used by the writer for list/footer indentation levels).

export class NestingStack<E> {
  private readonly stack: E[] = [];

  push(value: E): void {
    this.stack.push(value);
  }

  /** If the top is in `values`, pop and return it; else return undefined. */
  popIfIn(values: readonly E[]): E | undefined {
    if (this.stack.length === 0 || !values.includes(this.peek()!)) return undefined;
    return this.stack.pop();
  }

  /** If the stack contains `value`, pop it and everything above it; else nothing. */
  popUntil(value: E): void {
    if (!this.stack.includes(value)) return;
    let popped: E;
    do {
      popped = this.stack.pop()!;
    } while (popped !== value);
  }

  contains(value: E): boolean {
    return this.stack.includes(value);
  }

  containsAny(values: readonly E[]): boolean {
    return this.stack.some(e => values.includes(e));
  }

  isEmpty(): boolean {
    return this.stack.length === 0;
  }

  private peek(): E | undefined {
    return this.stack[this.stack.length - 1];
  }
}

/** Integer nesting stack tracking a running total (list/footer indent levels). */
export class IntNestingStack {
  private readonly stack: number[] = [];
  private _total = 0;

  total(): number {
    return this._total;
  }

  push(value = 1): void {
    this.stack.push(value);
    this._total += value;
  }

  popIfNotEmpty(): void {
    if (this.stack.length > 0) this._total -= this.stack.pop()!;
  }

  isEmpty(): boolean {
    return this.stack.length === 0;
  }

  reset(): void {
    this.stack.length = 0;
    this._total = 0;
  }
}

File: togo/internal/format/javadoc/nesting_stack.go
// Port of src/format/javadoc/nesting-stack.ts.

package javadoc

import "slices"

// nestingStack is a generic nesting stack (lexer HTML/code/table contexts).
type nestingStack[E comparable] struct {
	stack []E
}

func (s *nestingStack[E]) push(value E) { s.stack = append(s.stack, value) }

func (s *nestingStack[E]) popIfIn(values []E) (E, bool) {
	var zero E
	if len(s.stack) == 0 || !slices.Contains(values, s.stack[len(s.stack)-1]) {
		return zero, false
	}
	v := s.stack[len(s.stack)-1]
	s.stack = s.stack[:len(s.stack)-1]
	return v, true
}

// popUntil: if the stack contains value, pop it and everything above it.
func (s *nestingStack[E]) popUntil(value E) {
	if !slices.Contains(s.stack, value) {
		return
	}
	for {
		v := s.stack[len(s.stack)-1]
		s.stack = s.stack[:len(s.stack)-1]
		if v == value {
			return
		}
	}
}

func (s *nestingStack[E]) containsAny(values []E) bool {
	for _, e := range s.stack {
		if slices.Contains(values, e) {
			return true
		}
	}
	return false
}

func (s *nestingStack[E]) isEmpty() bool { return len(s.stack) == 0 }

// intNestingStack tracks a running total (writer list/footer indent levels).
type intNestingStack struct {
	stack []int
	total int
}

func (s *intNestingStack) push(value int) {
	s.stack = append(s.stack, value)
	s.total += value
}

func (s *intNestingStack) popIfNotEmpty() {
	if n := len(s.stack); n > 0 {
		s.total -= s.stack[n-1]
		s.stack = s.stack[:n-1]
	}
}

func (s *intNestingStack) isEmpty() bool { return len(s.stack) == 0 }

func (s *intNestingStack) reset() {
	s.stack = s.stack[:0]
	s.total = 0
}
